src/ts/components/TitleBar.ts

```diff
@@ -10,84 +12,101 @@ import { PlayerAPI } from 'bitmovin-player';
       cssClass: 'ui-titlebar',
       hidden: true,
-      components: [
-        new MetadataLabel({ content: MetadataLabelContent.Title }),
-        new MetadataLabel({ content: MetadataLabelContent.Description }),
-      ],
       keepHiddenWithoutMetadata: false,
+      components: hasCustomComponents
+        ? config.components
+        : [
+            new MetadataLabel({ content: MetadataLabelContent.Title }),
+            new MetadataLabel({ content: MetadataLabelContent.Description }),
+          ],
     }, <TitleBarConfig>this.config);
   }
 
   configure(player: PlayerAPI, uimanager: UIInstanceManager): void {
     super.configure(player, uimanager);
 
-    let config = this.getConfig();
+    const config = this.getConfig();
     let shouldBeShown = !this.isHidden();
-    let hasMetadataText = true; // Flag to track if any metadata label contains text
-
-    let checkMetadataTextAndUpdateVisibility = () => {
-      hasMetadataText = false;
-
-      // Iterate through metadata labels and check if at least one of them contains text
-      for (let component of this.getComponents()) {
-        if (component instanceof MetadataLabel) {
-          if (!component.isEmpty()) {
-            hasMetadataText = true;
-            break;
-          }
-        }
-      }
+    let hasVisibleText = true;
+
+    const checkTextAndUpdateVisibility = () => {
+      hasVisibleText = this.scanLabelsForText(this.getComponents());
 
       if (this.isShown()) {
-        // Hide a visible titlebar if it does not contain any text and the hidden flag is set
-        if (config.keepHiddenWithoutMetadata && !hasMetadataText) {
+        if (config.keepHiddenWithoutMetadata && !hasVisibleText) {
           this.hide();
         }
       } else if (shouldBeShown) {
-        // Show a hidden titlebar if it should actually be shown
-        this.show();
+        if (!config.keepHiddenWithoutMetadata || hasVisibleText) {
+          this.show();
+        }
       }
     };
 
-    // Listen to text change events to update the hasMetadataText flag when the metadata dynamically changes
-    for (let component of this.getComponents()) {
-      if (component instanceof MetadataLabel) {
-        component.onTextChanged.subscribe(checkMetadataTextAndUpdateVisibility);
-      }
-    }
+    // Subscribe to any label-like component's text changes
+    this.subscribeToTextChanges(this.getComponents(), checkTextAndUpdateVisibility);
 
     uimanager.onControlsShow.subscribe(() => {
       shouldBeShown = true;
-      if (!(config.keepHiddenWithoutMetadata && !hasMetadataText)) {
+      if (!config.keepHiddenWithoutMetadata || hasVisibleText) {
         this.show();
       }
     });
+
     uimanager.onControlsHide.subscribe(() => {
       shouldBeShown = false;
       this.hide();
     });
 
-    // init
-    checkMetadataTextAndUpdateVisibility();
+    checkTextAndUpdateVisibility();
+  }
+
+  private scanLabelsForText(components: Component<ComponentConfig>[]): boolean {
```

Comment:
I think we should improve the naming of this. The name `scanlabelsForText` with a return value of type Boolean doesn't really indicate what the `Boolean` represents. Is it now `true` if there are labels with text or is it true if there are no labels with text.

+ in combination with the previous comment:
```suggestion
  private hasNonEmptyComponents(components: Component<ComponentConfig>[]): boolean {
```

Reply:
Address in a36a15f1b4bd67856658a4a086a25cee728198f7